File: Clients/Node/Admin/src/services/AudioConfigurationService.ts
import { FetchBaseApiClient } from '../client/FetchBaseApiClient';
import type { AudioProviderSettings } from '../models/common-types';
import type {
  AudioProviderConfigRequest,
  AudioProviderConfigDto,
  AudioCostConfigRequest,
  AudioCostConfigDto,
  AudioUsageDto,
  AudioUsageSummaryDto,
  AudioUsageFilters,
  AudioUsageSummaryFilters,
  RealtimeSessionDto,
  AudioProviderTestResult,
} from '../models/audioConfiguration';
import type { PagedResponse } from '../models/common';
import {
  validateAudioProviderRequest,
  validateAudioCostConfigRequest,
  validateAudioUsageFilters,
} from '../models/audioConfiguration';

/**
 * Service for managing audio provider configurations, cost settings, and usage analytics
 */
export class AudioConfigurationService {
  private static readonly PROVIDERS_ENDPOINT = '/api/admin/audio/providers';
  private static readonly COSTS_ENDPOINT = '/api/admin/audio/costs';
  private static readonly USAGE_ENDPOINT = '/api/admin/audio/usage';
  private static readonly SESSIONS_ENDPOINT = '/api/admin/audio/sessions';

  constructor(private readonly client: FetchBaseApiClient) {}

  // #region Provider Configuration

  /**
   * Creates a new audio provider configuration
   */
  async createProvider(request: AudioProviderConfigRequest): Promise<AudioProviderConfigDto> {
    validateAudioProviderRequest(request);
    
    return this.client['post']<AudioProviderConfigDto>(
      AudioConfigurationService.PROVIDERS_ENDPOINT,
      request
    );
  }

  /**
   * Gets all audio provider configurations
   */
  async getProviders(): Promise<AudioProviderConfigDto[]> {
    return this.client['get']<AudioProviderConfigDto[]>(
      AudioConfigurationService.PROVIDERS_ENDPOINT
    );
  }

  /**
   * Gets enabled audio providers for a specific operation type
   */
  async getEnabledProviders(operationType: string): Promise<AudioProviderConfigDto[]> {
    if (!operationType || operationType.trim().length === 0) {
      throw new Error('Operation type is required');
    }

    const endpoint = `${AudioConfigurationService.PROVIDERS_ENDPOINT}/enabled/${encodeURIComponent(operationType)}`;
    
    return this.client['get']<AudioProviderConfigDto[]>(endpoint);
  }

  /**
   * Gets a specific audio provider configuration by ID
   */
  async getProvider(providerId: string): Promise<AudioProviderConfigDto> {
    if (!providerId || providerId.trim().length === 0) {
      throw new Error('Provider ID is required');
    }

    const endpoint = `${AudioConfigurationService.PROVIDERS_ENDPOINT}/${encodeURIComponent(providerId)}`;
    
    return this.client['get']<AudioProviderConfigDto>(endpoint);
  }

  /**
   * Updates an existing audio provider configuration
   */
  async updateProvider(
    providerId: string,
    request: AudioProviderConfigRequest
  ): Promise<AudioProviderConfigDto> {
    if (!providerId || providerId.trim().length === 0) {
      throw new Error('Provider ID is required');
    }

    validateAudioProviderRequest(request);

    const endpoint = `${AudioConfigurationService.PROVIDERS_ENDPOINT}/${encodeURIComponent(providerId)}`;
    
    return this.client['put']<AudioProviderConfigDto>(
      endpoint,
      request
    );
  }

  /**
   * Deletes an audio provider configuration
   */
  async deleteProvider(providerId: string): Promise<void> {
    if (!providerId || providerId.trim().length === 0) {
      throw new Error('Provider ID is required');
    }

    const endpoint = `${AudioConfigurationService.PROVIDERS_ENDPOINT}/${encodeURIComponent(providerId)}`;
    
    await this.client['delete']<void>(endpoint);
  }

  /**
   * Tests the connectivity and configuration of an audio provider
   */
  async testProvider(providerId: string): Promise<AudioProviderTestResult> {
    if (!providerId || providerId.trim().length === 0) {
      throw new Error('Provider ID is required');
    }

    const endpoint = `${AudioConfigurationService.PROVIDERS_ENDPOINT}/${encodeURIComponent(providerId)}/test`;
    
    return this.client['post']<AudioProviderTestResult>(
      endpoint,
      {}
    );
  }

  // #endregion

  // #region Cost Configuration

  /**
   * Creates a new audio cost configuration
   */
  async createCostConfig(request: AudioCostConfigRequest): Promise<AudioCostConfigDto> {
    validateAudioCostConfigRequest(request);
    
    return this.client['post']<AudioCostConfigDto>(
      AudioConfigurationService.COSTS_ENDPOINT,
      request
    );
  }

  /**
   * Gets all audio cost configurations
   */
  async getCostConfigs(): Promise<AudioCostConfigDto[]> {
    return this.client['get']<AudioCostConfigDto[]>(
      AudioConfigurationService.COSTS_ENDPOINT
    );
  }

  /**
   * Gets a specific audio cost configuration by ID
   */
  async getCostConfig(configId: string): Promise<AudioCostConfigDto> {
    if (!configId || configId.trim().length === 0) {
      throw new Error('Cost configuration ID is required');
    }

    const endpoint = `${AudioConfigurationService.COSTS_ENDPOINT}/${encodeURIComponent(configId)}`;
    
    return this.client['get']<AudioCostConfigDto>(endpoint);
  }

  /**
   * Updates an existing audio cost configuration
   */
  async updateCostConfig(
    configId: string,
    request: AudioCostConfigRequest
  ): Promise<AudioCostConfigDto> {
    if (!configId || configId.trim().length === 0) {
      throw new Error('Cost configuration ID is required');
    }

    validateAudioCostConfigRequest(request);

    const endpoint = `${AudioConfigurationService.COSTS_ENDPOINT}/${encodeURIComponent(configId)}`;
    
    return this.client['put']<AudioCostConfigDto>(
      endpoint,
      request
    );
  }

  /**
   * Deletes an audio cost configuration
   */
  async deleteCostConfig(configId: string): Promise<void> {
    if (!configId || configId.trim().length === 0) {
      throw new Error('Cost configuration ID is required');
    }

    const endpoint = `${AudioConfigurationService.COSTS_ENDPOINT}/${encodeURIComponent(configId)}`;
    
    await this.client['delete']<void>(endpoint);
  }

  // #endregion

  // #region Usage Analytics

  /**
   * Gets audio usage data with optional filtering
   */
  async getUsage(filters?: AudioUsageFilters): Promise<PagedResponse<AudioUsageDto>> {
    if (filters) {
      validateAudioUsageFilters(filters);
    }

    const queryParams: string[] = [];
    
    if (filters?.startDate) {
      queryParams.push(`startDate=${encodeURIComponent(filters.startDate)}`);
    }
    
    if (filters?.endDate) {
      queryParams.push(`endDate=${encodeURIComponent(filters.endDate)}`);
    }
    
    if (filters?.virtualKey) {
      queryParams.push(`virtualKey=${encodeURIComponent(filters.virtualKey)}`);
    }
    
    if (filters?.provider) {
      queryParams.push(`provider=${encodeURIComponent(filters.provider)}`);
    }
    
    if (filters?.operationType) {
      queryParams.push(`operationType=${encodeURIComponent(filters.operationType)}`);
    }
    
    if (filters?.page !== undefined) {
      queryParams.push(`page=${filters.page}`);
    }
    
    if (filters?.pageSize !== undefined) {
      queryParams.push(`pageSize=${filters.pageSize}`);
    }

    const endpoint = queryParams.length > 0 
      ? `${AudioConfigurationService.USAGE_ENDPOINT}?${queryParams.join('&')}`
      : AudioConfigurationService.USAGE_ENDPOINT;
    
    return this.client['get']<PagedResponse<AudioUsageDto>>(endpoint);
  }

  /**
   * Gets audio usage summary for a date range
   */
  async getUsageSummary(filters: AudioUsageSummaryFilters): Promise<AudioUsageSummaryDto> {
    if (!filters.startDate || !filters.endDate) {
      throw new Error('Start date and end date are required for usage summary');
    }

    if (new Date(filters.startDate) >= new Date(filters.endDate)) {
      throw new Error('Start date must be before end date');
    }

    const queryParams: string[] = [
      `startDate=${encodeURIComponent(filters.startDate)}`,
      `endDate=${encodeURIComponent(filters.endDate)}`,
    ];
    
    if (filters.virtualKey) {
      queryParams.push(`virtualKey=${encodeURIComponent(filters.virtualKey)}`);
    }
    
    if (filters.provider) {
      queryParams.push(`provider=${encodeURIComponent(filters.provider)}`);
    }
    
    if (filters.operationType) {
      queryParams.push(`operationType=${encodeURIComponent(filters.operationType)}`);
    }

    const endpoint = `${AudioConfigurationService.USAGE_ENDPOINT}/summary?${queryParams.join('&')}`;
    
    return this.client['get']<AudioUsageSummaryDto>(endpoint);
  }

  // #endregion

  // #region Real-time Sessions

  /**
   * Gets all active real-time audio sessions
   */
  async getActiveSessions(): Promise<RealtimeSessionDto[]> {
    return this.client['get']<RealtimeSessionDto[]>(
      AudioConfigurationService.SESSIONS_ENDPOINT
    );
  }

  /**
   * Gets a specific real-time session by ID
   */
  async getSession(sessionId: string): Promise<RealtimeSessionDto> {
    if (!sessionId || sessionId.trim().length === 0) {
      throw new Error('Session ID is required');
    }

    const endpoint = `${AudioConfigurationService.SESSIONS_ENDPOINT}/${encodeURIComponent(sessionId)}`;
    
    return this.client['get']<RealtimeSessionDto>(endpoint);
  }

  /**
   * Terminates an active real-time audio session
   */
  async terminateSession(sessionId: string): Promise<{ success: boolean; sessionId: string; message?: string }> {
    if (!sessionId || sessionId.trim().length === 0) {
      throw new Error('Session ID is required');
    }

    const endpoint = `${AudioConfigurationService.SESSIONS_ENDPOINT}/${encodeURIComponent(sessionId)}/terminate`;
    
    try {
      const response = await this.client['post']<{ success: boolean; message?: string }>(endpoint, {});
      return {
        success: response.success,
        sessionId,
        message: response.message,
      };
    } catch (error) {
      // If the session is already terminated or doesn't exist, handle gracefully
      if (error && typeof error === 'object' && 'status' in error) {
        if (error.status === 404) {
          throw new Error('Session not found or already terminated');
        } else if (error.status === 409) {
          throw new Error('Session is already terminated');
        }
      }
      throw error;
    }
  }

  // #endregion

}

/**
 * Helper functions for audio configuration management
 */
export const AudioConfigurationHelpers = {
  /**
   * Creates a basic audio provider configuration request
   */
  createProviderRequest(
    name: string,
    baseUrl: string,
    apiKey: string,
    options?: {
      isEnabled?: boolean;
      supportedOperations?: string[];
      priority?: number;
      timeoutSeconds?: number;
      settings?: AudioProviderSettings;
    }
  ): AudioProviderConfigRequest {
    return {
      name,
      baseUrl,
      apiKey,
      isEnabled: options?.isEnabled ?? true,
      supportedOperations: options?.supportedOperations ?? [],
      priority: options?.priority ?? 1,
      timeoutSeconds: options?.timeoutSeconds ?? 30,
      settings: options?.settings,
    };
  },

  /**
   * Creates a basic audio cost configuration request
   */
  createCostConfigRequest(
    providerId: string,
    operationType: string,
    costPerUnit: number,
    unitType: string,
    options?: {
      modelName?: string;
      currency?: string;
      isActive?: boolean;
      effectiveFrom?: string;
      effectiveTo?: string;
    }
  ): AudioCostConfigRequest {
    return {
      providerId,
      operationType,
      costPerUnit,
      unitType,
      modelName: options?.modelName,
      currency: options?.currency ?? 'USD',
      isActive: options?.isActive ?? true,
      effectiveFrom: options?.effectiveFrom,
      effectiveTo: options?.effectiveTo,
    };
  },

  /**
   * Creates audio usage filters with sensible defaults
   */
  createUsageFilters(options?: {
    startDate?: string;
    endDate?: string;
    virtualKey?: string;
    provider?: string;
    operationType?: string;
    page?: number;
    pageSize?: number;
  }): AudioUsageFilters {
    return {
      startDate: options?.startDate,
      endDate: options?.endDate,
      virtualKey: options?.virtualKey,
      provider: options?.provider,
      operationType: options?.operationType,
      page: options?.page ?? 1,
      pageSize: options?.pageSize ?? 50,
    };
  },

  /**
   * Creates date range for common periods
   */
  createDateRange(period: 'today' | 'yesterday' | 'last7days' | 'last30days' | 'thisMonth'): {
    startDate: string;
    endDate: string;
  } {
    const now = new Date();
    const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    
    switch (period) {
      case 'today':
        return {
          startDate: today.toISOString(),
          endDate: now.toISOString(),
        };
      
      case 'yesterday': {
        const yesterday = new Date(today);
        yesterday.setDate(yesterday.getDate() - 1);
        return {
          startDate: yesterday.toISOString(),
          endDate: today.toISOString(),
        };
      }
      
      case 'last7days': {
        const sevenDaysAgo = new Date(today);
        sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7);
        return {
          startDate: sevenDaysAgo.toISOString(),
          endDate: now.toISOString(),
        };
      }
      
      case 'last30days': {
        const thirtyDaysAgo = new Date(today);
        thirtyDaysAgo.setDate(thirtyDaysAgo.getDate() - 30);
        return {
          startDate: thirtyDaysAgo.toISOString(),
          endDate: now.toISOString(),
        };
      }
      
      case 'thisMonth': {
        const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
        return {
          startDate: monthStart.toISOString(),
          endDate: now.toISOString(),
        };
      }
      
      default:
        throw new Error(`Unknown period: ${period}`);
    }
  },
};

File: archive/Clients.DotNet/ConduitLLM.AdminClient/src/Services/AudioConfigurationService.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Caching.Memory;
using ConduitLLM.AdminClient.Client;
using ConduitLLM.AdminClient.Exceptions;
using ConduitLLM.AdminClient.Models;
using ConduitLLM.AdminClient.Utils;

namespace ConduitLLM.AdminClient.Services
{
    /// <summary>
    /// Service for managing audio provider configurations, cost settings, and usage analytics.
    /// </summary>
    public class AudioConfigurationService : BaseApiClient
    {
        private const string ProvidersEndpoint = "/api/admin/audio/providers";
        private const string CostsEndpoint = "/api/admin/audio/costs";
        private const string UsageEndpoint = "/api/admin/audio/usage";
        private const string SessionsEndpoint = "/api/admin/audio/sessions";

        /// <summary>
        /// Initializes a new instance of the <see cref="AudioConfigurationService"/> class.
        /// </summary>
        /// <param name="httpClient">The HTTP client to use for requests.</param>
        /// <param name="configuration">The client configuration.</param>
        /// <param name="logger">Optional logger instance.</param>
        /// <param name="cache">Optional memory cache instance.</param>
        public AudioConfigurationService(
            HttpClient httpClient,
            ConduitAdminClientConfiguration configuration,
            ILogger<AudioConfigurationService>? logger = null,
            IMemoryCache? cache = null)
            : base(httpClient, configuration, logger, cache)
        {
        }

        #region Provider Configuration

        /// <summary>
        /// Creates a new audio provider configuration.
        /// </summary>
        /// <param name="request">The provider configuration request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The created provider configuration.</returns>
        /// <exception cref="ValidationException">Thrown when the request is invalid.</exception>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<AudioProviderConfigDto> CreateProviderAsync(
            AudioProviderConfigRequest request,
            CancellationToken cancellationToken = default)
        {
            try
            {
                ValidateProviderRequest(request);
                
                // Creating audio provider configuration

                var response = await PostAsync<AudioProviderConfigDto>(
                    ProvidersEndpoint,
                    request,
                    cancellationToken);

                // Successfully created audio provider
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        /// <summary>
        /// Gets all audio provider configurations.
        /// </summary>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>List of provider configurations.</returns>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<List<AudioProviderConfigDto>> GetProvidersAsync(
            CancellationToken cancellationToken = default)
        {
            try
            {
                // Retrieving all audio provider configurations

                var response = await GetAsync<List<AudioProviderConfigDto>>(
                    ProvidersEndpoint,
                    cancellationToken);

                // Retrieved audio provider configurations
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        /// <summary>
        /// Tests the connectivity and configuration of an audio provider.
        /// </summary>
        /// <param name="providerId">The provider ID to test.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The test result.</returns>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<AudioProviderTestResult> TestProviderAsync(
            string providerId,
            CancellationToken cancellationToken = default)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(providerId))
                    throw new ValidationException("Provider ID is required", "providerId");

                // Testing audio provider

                var endpoint = $"{ProvidersEndpoint}/{Uri.EscapeDataString(providerId)}/test";
                var response = await PostAsync<AudioProviderTestResult>(
                    endpoint,
                    new { },
                    cancellationToken);

                // Provider test completed
                
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        #endregion

        #region Cost Configuration

        /// <summary>
        /// Creates a new audio cost configuration.
        /// </summary>
        /// <param name="request">The cost configuration request.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The created cost configuration.</returns>
        /// <exception cref="ValidationException">Thrown when the request is invalid.</exception>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<AudioCostConfigDto> CreateCostConfigAsync(
            AudioCostConfigRequest request,
            CancellationToken cancellationToken = default)
        {
            try
            {
                ValidateCostConfigRequest(request);
                
                // Creating audio cost configuration

                var response = await PostAsync<AudioCostConfigDto>(
                    CostsEndpoint,
                    request,
                    cancellationToken);

                // Successfully created audio cost configuration
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        /// <summary>
        /// Gets all audio cost configurations.
        /// </summary>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>List of cost configurations.</returns>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<List<AudioCostConfigDto>> GetCostConfigsAsync(
            CancellationToken cancellationToken = default)
        {
            try
            {
                // Retrieving all audio cost configurations

                var response = await GetAsync<List<AudioCostConfigDto>>(
                    CostsEndpoint,
                    cancellationToken);

                // Retrieved audio cost configurations
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        #endregion

        #region Usage Analytics

        /// <summary>
        /// Gets audio usage data with optional filtering.
        /// </summary>
        /// <param name="startDate">Optional start date filter.</param>
        /// <param name="endDate">Optional end date filter.</param>
        /// <param name="virtualKey">Optional virtual key filter.</param>
        /// <param name="provider">Optional provider filter.</param>
        /// <param name="operationType">Optional operation type filter.</param>
        /// <param name="page">Page number for pagination (1-based).</param>
        /// <param name="pageSize">Number of items per page.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Paginated list of usage data.</returns>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<PagedResponse<AudioUsageDto>> GetUsageAsync(
            DateTime? startDate = null,
            DateTime? endDate = null,
            string? virtualKey = null,
            string? provider = null,
            string? operationType = null,
            int page = 1,
            int pageSize = 50,
            CancellationToken cancellationToken = default)
        {
            try
            {
                // Retrieving audio usage data with filters

                var queryParams = new List<string>();
                
                if (startDate.HasValue)
                    queryParams.Add($"startDate={startDate.Value:yyyy-MM-ddTHH:mm:ssZ}");
                
                if (endDate.HasValue)
                    queryParams.Add($"endDate={endDate.Value:yyyy-MM-ddTHH:mm:ssZ}");
                
                if (!string.IsNullOrWhiteSpace(virtualKey))
                    queryParams.Add($"virtualKey={Uri.EscapeDataString(virtualKey)}");
                
                if (!string.IsNullOrWhiteSpace(provider))
                    queryParams.Add($"provider={Uri.EscapeDataString(provider)}");
                
                if (!string.IsNullOrWhiteSpace(operationType))
                    queryParams.Add($"operationType={Uri.EscapeDataString(operationType)}");
                
                queryParams.Add($"page={page}");
                queryParams.Add($"pageSize={pageSize}");

                var endpoint = $"{UsageEndpoint}?{string.Join("&", queryParams)}";
                var response = await GetAsync<PagedResponse<AudioUsageDto>>(
                    endpoint,
                    cancellationToken);

                // Retrieved usage records
                
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        /// <summary>
        /// Gets audio usage summary for a date range.
        /// </summary>
        /// <param name="startDate">Start date for the summary.</param>
        /// <param name="endDate">End date for the summary.</param>
        /// <param name="virtualKey">Optional virtual key filter.</param>
        /// <param name="provider">Optional provider filter.</param>
        /// <param name="operationType">Optional operation type filter.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Usage summary data.</returns>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<AudioUsageSummaryDto> GetUsageSummaryAsync(
            DateTime startDate,
            DateTime endDate,
            string? virtualKey = null,
            string? provider = null,
            string? operationType = null,
            CancellationToken cancellationToken = default)
        {
            try
            {
                // Retrieving audio usage summary

                var queryParams = new List<string>
                {
                    $"startDate={startDate:yyyy-MM-ddTHH:mm:ssZ}",
                    $"endDate={endDate:yyyy-MM-ddTHH:mm:ssZ}"
                };
                
                if (!string.IsNullOrWhiteSpace(virtualKey))
                    queryParams.Add($"virtualKey={Uri.EscapeDataString(virtualKey)}");
                
                if (!string.IsNullOrWhiteSpace(provider))
                    queryParams.Add($"provider={Uri.EscapeDataString(provider)}");
                
                if (!string.IsNullOrWhiteSpace(operationType))
                    queryParams.Add($"operationType={Uri.EscapeDataString(operationType)}");

                var endpoint = $"{UsageEndpoint}/summary?{string.Join("&", queryParams)}";
                var response = await GetAsync<AudioUsageSummaryDto>(
                    endpoint,
                    cancellationToken);

                // Retrieved usage summary
                
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        #endregion

        #region Real-time Sessions

        /// <summary>
        /// Gets all active real-time audio sessions.
        /// </summary>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>List of active sessions.</returns>
        /// <exception cref="ConduitAdminException">Thrown when the API request fails.</exception>
        public async Task<List<RealtimeSessionDto>> GetActiveSessionsAsync(
            CancellationToken cancellationToken = default)
        {
            try
            {
                // Retrieving active audio sessions

                var response = await GetAsync<List<RealtimeSessionDto>>(
                    SessionsEndpoint,
                    cancellationToken);

                // Retrieved active audio sessions
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitAdminException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        #endregion

        #region Validation

        private static void ValidateProviderRequest(AudioProviderConfigRequest request)
        {
            if (request == null)
                throw new ValidationException("Provider configuration request cannot be null", "request");

            if (string.IsNullOrWhiteSpace(request.Name))
                throw new ValidationException("Provider name is required", "name");

            if (string.IsNullOrWhiteSpace(request.BaseUrl))
                throw new ValidationException("Base URL is required", "baseUrl");

            if (string.IsNullOrWhiteSpace(request.ApiKey))
                throw new ValidationException("API key is required", "apiKey");

            if (!Uri.TryCreate(request.BaseUrl, UriKind.Absolute, out var uri) ||
                (uri.Scheme != "http" && uri.Scheme != "https"))
            {
                throw new ValidationException("Base URL must be a valid HTTP or HTTPS URL", "baseUrl");
            }

            if (request.TimeoutSeconds <= 0 || request.TimeoutSeconds > 300)
                throw new ValidationException("Timeout must be between 1 and 300 seconds", "timeoutSeconds");

            if (request.Priority < 1)
                throw new ValidationException("Priority must be at least 1", "priority");
        }

        private static void ValidateCostConfigRequest(AudioCostConfigRequest request)
        {
            if (request == null)
                throw new ValidationException("Cost configuration request cannot be null", "request");

            if (string.IsNullOrWhiteSpace(request.ProviderId))
                throw new ValidationException("Provider ID is required", "providerId");

            if (string.IsNullOrWhiteSpace(request.OperationType))
                throw new ValidationException("Operation type is required", "operationType");

            if (string.IsNullOrWhiteSpace(request.UnitType))
                throw new ValidationException("Unit type is required", "unitType");

            if (request.CostPerUnit < 0)
                throw new ValidationException("Cost per unit cannot be negative", "costPerUnit");

            if (request.EffectiveFrom.HasValue && request.EffectiveTo.HasValue &&
                request.EffectiveFrom >= request.EffectiveTo)
            {
                throw new ValidationException("Effective from date must be before effective to date", "effectiveFrom");
            }
        }

        #endregion
    }
}